clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -256,12 +266,28 @@ private long listOffset(TopicPartition partition, long timestamp) {
      * Return the fetched records, empty the record buffer and update the consumed position.
      *
      * @return The fetched records per partition
+     * @throws OffsetOutOfRangeException If there is OffsetOutOfRange error in fetchResponse
      */
     public Map<TopicPartition, List<ConsumerRecord<K, V>>> fetchedRecords() {
         if (this.subscriptions.partitionAssignmentNeeded()) {
             return Collections.emptyMap();
         } else {
             Map<TopicPartition, List<ConsumerRecord<K, V>>> drained = new HashMap<>();
+            Map<TopicPartition, Long> currentOutOfRangePartitions = new HashMap<>();
+
+            // filter offsetOutOfRangePartitions to retain only the fetchable partitions
+            for (Map.Entry<TopicPartition, Long> entry: this.offsetOutOfRangePartitions.entrySet()) {
+                if (!subscriptions.isFetchable(entry.getKey())) {
+                    log.debug("Ignoring fetched records for {} since it is no longer fetchable", entry.getKey());
+                    continue;
+                }
+                currentOutOfRangePartitions.put(entry.getKey(), entry.getValue());
+            }
+            this.offsetOutOfRangePartitions.clear();
+            if (!currentOutOfRangePartitions.isEmpty())
+                throw new OffsetOutOfRangeException(currentOutOfRangePartitions);
```

Comment:
Could we turn this block into a method? For example, throwIfOutofRange() or something like that.